File: tools/tal_support/as274_fc/F/t3.f
      program test3
c
c     Test vmove for treatment of singularities.
c
      real rand
      double precision d(7), rbar(21), thetab(7), x(7), y 
      double precision sserr, work(7),
     +       tol(7), rss(7), beta(7)
      integer ix, iy, iz, i, ier, vorder(7)
      logical lindep(7)
      common /randc/ ix, iy, iz

      ix = 777
      iy = 777
      iz = 777

c     Initialize orthogonal reduction.
      call clear(7, 21, d, rbar, thetab, sserr, ier)
      do 10 i = 1, 7
	vorder(i) = i
   10 continue

c     Generate data such that:
c             x2 = x4 - x5
c             x6 = x1 - x3
c             y  = x1 + x3 + x4 + x5 + x7

      do 30 i = 1, 12
	x(1) = rand()
	x(3) = rand()
	x(4) = rand()
	x(5) = rand()
	x(7) = rand()
	x(2) = x(4) - x(5)
	x(6) = x(1) - x(3)
	y = x(1) + x(3) + x(4) + x(5) + x(7)
	call includ(7, 21, 1.0d0, x, y, d, rbar, thetab, sserr, ier)
   30 continue
      write(*, *)'As output from includ:'
      write(*, 900) d, rbar, thetab
  900 format(' d:'/1x,7f11.6/
     +       ' rbar:'/1x,6f11.6/ 12x,5f11.6/ 23x,4f11.6/ 34x,3f11.6/
     +             45x,2f11.6/ 56x,f11.6/
     +       ' thetab:'/1x,7f11.6//)
c
c     Set up arrays TOL & RSS.
c
      call tolset(7, 21, d, rbar, tol, work, ier)
      call ss(7, d, thetab, sserr, rss, ier)
      write(*, 920) rss
  920 format(' RSS:'/ 1x, 7f11.5/)
c
c     Use SING to set near zeroes to zero.
c
      call sing(7, 21, d, rbar, thetab, sserr, tol, lindep, work, ier)
      write(*, *)'After being processed by sing:'
      write(*, *)'IER = ', ier
      write(*, 900) d, rbar, thetab
      call ss(7, d, thetab, sserr, rss, ier)
      write(*, 920) rss
c
c       Swap rows 4 & 5 and rows 6 & 7.
c
      call vmove(7, 21, vorder, d, rbar, thetab, rss, 4, 5, tol, ier)
      call vmove(7, 21, vorder, d, rbar, thetab, rss, 6, 7, tol, ier)
      write(*, *)'After interchange of variables 4 & 5, and 6 & 7'
      write(*, 900) d, rbar, thetab
      call regcf(7, 21, d, rbar, thetab, tol, beta, 7, ier)
      write(*, 910) beta
  910 format(' Regression coefficients:'/ 1x,7f11.6)
      call ss(7, d, thetab, sserr, rss, ier)
      write(*, 920) rss
      end

File: tools/tal_support/as274_fc/C/t3.c
#include <math.h>
#include <stdio.h>
#include "as274c.h"
#include "misc.h"

int main()
{
    double *d, *rbar, *thetab, *x, y, sserr, *tol, *rss, *beta;
    int i, error, *vorder, *lindep;

    d = dvector(0, 6);
    rbar = dvector(0, 20);
    thetab = dvector(0, 6);
    x = dvector(0, 6);
    tol = dvector(0, 6);
    rss = dvector(0, 6);
    beta = dvector(0, 6);
    vorder = ivector(0, 6);
    lindep = ivector(0, 6);

    randc.ix = 777;
    randc.iy = 777;
    randc.iz = 777;

    error = clear(7, 21, d, rbar, thetab, &sserr);
    if (error) printf("clear() returned %d\n", error);
    for (i=0; i<7; i++) vorder[i] = i;
    for (i=0; i<12; i++) {
        x[0] = ranwm();
        x[2] = ranwm();
        x[3] = ranwm();
        x[4] = ranwm();
        x[6] = ranwm();
        x[1] = x[3] - x[4];
        x[5] = x[0] - x[2];
        y = x[0] + x[2] + x[3] + x[4] + x[6];
        error = includ(7, 21, 1, x, y, d, rbar, thetab, &sserr);
        if (error) printf("includ() returned %d\n", error);
    }

#define PRINTTHREE \
    printf("\n"); \
    putdvec("d", d, 0, 6); \
    pr_utdm_v(rbar, 6, 12, 6); \
    putdvec("thetab", thetab, 0, 6); \
    printf("\n")

    PRINTTHREE;

    error = tolset(7, 21, d, rbar, tol);
    if (error) printf("tolset() returned %d\n", error);
    error = ss(7, d, thetab, &sserr, rss);
    if (error) printf("ss() returned %d\n", error);
    putdvec("RSS", rss, 0, 6);

    error = sing(7, 21, d, rbar, thetab, &sserr, tol, lindep);
    if (error) printf("sing() returned %d\n", error);
    printf("After processing by sing():\n");
    PRINTTHREE;
    error = ss(7, d, thetab, &sserr, rss);
    if (error) printf("ss() returned %d\n", error);
    putdvec("RSS", rss, 0, 6);

    /* Swap rows 4 and 5, and 6 and 7 */
    error = vmove(7, 21, vorder, d, rbar, thetab, rss, 4, 5, tol);
    error = vmove(7, 21, vorder, d, rbar, thetab, rss, 6, 7, tol);
    PRINTTHREE;
    error = regcf(7, 21, d, rbar, thetab, tol, beta, 7);
    putdvec("Beta:", beta, 0, 6);
    error = ss(7, d, thetab, &sserr, rss);
    if (error) printf("ss() returned %d\n", error);
    putdvec("RSS", rss, 0, 6);
    return 0;
}
